First version

g-arjones opened the pull request:
This uses some code from the daemon. It's an idea I had for quite a while and finally took the time to do it. The goal is to make testing pull requests easier for team members.

Use it with `autoproj overrides "https://github.com/rock-core/autoproj-overrides-generator/pull/1"` to get overrides for the given pull request based on the "depends on" pattern.

g-arjones committed:
Initial commit

g-arjones committed:
Add tests

g-arjones committed:
Use github access tokens rather than user and password

g-arjones committed:
Create overrides dir if it does not exist

g-arjones committed:
Add README

doudou commented:
Yeah. I had that on my wishlist for a long time now.

I'd suggest to rather create a `github` plugin that would have a subcommand for that. I have a few idea of other github-related tools that can be implemented. For instance, `push-init` which would create the first version of a repo (create master, initial branches, pull requests on package and package set(s) - all with the right dependencies and based on where the autoproj config expects it to be ;-)). Or a pull-request template tool that would ask for `Depends on` on dependencies that have non-mainlined changes (:P). 

From a usability standpoint,
- I would have expected to only give the package name (or something that matches a package name) instead of a whole URL. The tool can resolve the name-to-URL
- I would have expected to at least have the option to provide the output file (or, rather, a way to output to standard out)

None of this is a blocker for me, just suggestions ...

g-arjones commented:
> I'd suggest to rather create a github plugin

Sounds good to me. I like all the subcommand ideas. :+1: 

> I would have expected to only give the package name

I can do that in addition to the full URL because I would like to be able to test i.e: changing a package to a forked version of it.

> I would have expected to at least have the option to provide the output file

Sure! :+1: 

> a way to output to standard out

I am not sure there's a good way to do that because in order to resolve packages the plugin might need to update package set branches (when you are testing a pull request in a package that is not defined yet). What I do is look for package sets in the list of dependencies, save a partial overrides file, restart and repeat until there's no package set that requires updating. I'll be glad to change this if you have a better idea

doudou commented:
> I am not sure there's a good way to do that because in order to resolve packages the plugin might need to update package set branches (when you are testing a pull request in a package that is not defined yet). What I do is look for package sets in the list of dependencies, save a partial overrides file, restart and repeat until there's no package set that requires updating. I'll be glad to change this if you have a better idea

I didn't think about that. Sounds reasonable. This is actually an improvement over the daemon's behavior. Adding a new package is a bit of a pain right now.

doudou commented:
> because I would like to be able to test i.e: changing a package to a forked version of it.

Wow. I like the sounds of that.

g-arjones commented:
> This is actually an improvement over the daemon's behavior

I agree and I believe it could be implemented in the daemon as well but it would be much trickier. You have to make sure you don't process any events while you're updating package sets and generating overrides (which may involve multiple restarts) and you gotta make sure it gets back to mainline once you are done and even (or especially) if something fails.

To be honest, writing plugins for autoproj would be much easier if we didn't have work around all the global state it and autobuild keep. I think this is the greatest design improvent they could see right now. Volunteer? :P

g-arjones committed:
Rename plugin to github

g-arjones committed:
Allow defining overrides filename

g-arjones commented:
Done...

1. Renamed to autoproj-github
2. Added `--output` to set the output filename

Did not add the option to resolve urls from package name because you still need to know the PR number (most likely having to open the browser anyway) so it didn't seem like much of an improvement.
